src/Neyro.AppMetrics.Extensions.EventCountersCollector/Neyro.AppMetrics.Extensions.EventCountersCollector/CounterPayload.cs

```diff
@@ -40,14 +50,20 @@ public CounterPayload(IDictionary<string, object> payloadFields)
                 }
             }
 
-            Metadata = ExtractTagsFromMetadata(payloadFields);
+            if (useMetadata)
+            {
+                ExtractTagsFromMetadata(payloadFields);
+                // setting the key from original metadata so avoid allocation if we join
+                // the values in the dictionary when creating cache key
+                payloadFields.TryGetValue("Metadata", out object metadataSource);
+                Key = $"{Name}{metadataSource as string}";
+            }
```

Comment:
We called Dictionary.ContainsKey and Dictionary's indexer in ExtractTagsFromMetadata. And now call Dictionary.TryGetValue here again, which is equivalent to call Dictionary.ContainsKey and Dictionary's indexer.
I suggest:
1) Change this fragment
```
if (useMetadata && payloadFields.TryGetValue("Metadata", out object metadataSource))
{
    var payloadMetadata = metadataSource as string;
    if(string.IsNullOrWhitespace(payloadMetadata)) 
       return;
    ExtractTagsFromMetadata(payloadFields);
    Key = Name + payloadMetadata;
}
```
2) change method ExtractTagsFromMetadata
```
private void ExtractTagsFromMetadata(string payloadMetadata)
{
   var metaKVs = payloadMetadata.Split(",");
   foreach(var strKv in metaKVs)
   {
      var kv = strKv.Split(":");
      if (kv.Length != 2)
          continue;
      Metadata.Add(kv[0], kv[1]);
    }
}
```

Reply:
Good call @neyromant I've reduced the allocation and changed the code a bit more, much better now.